### DISCUSSION TITLE
Replacing colors in PNG withou changing image format

### DISCUSSION THREAD

**Discussion by Oblomov** [ImageMagick/ImageMagick#6823] [Help] [answered]
### ImageMagick version

7.1.1-20

### Operating system, version and so on

Debian unstable

### Description

Hello,

I'm looking for a way to replace a color (exact match) in a PNG while leaving everything else unchanged. However, if I try to do

    magick convert original.png -fill $newcolor -opaque $oldcolor newfile.png

the output file format changes: the original is DirectClass, the output PseudoClass ; original has 1-bit alpha channel, the output does not. Adding -type TrueColorAlpha does not help preserve these properties. Is there anything I can do to maintain the same exact color space as the original file?

**Comment by snibgo** [comment]
If you showed a sample input and output, that would help us understand.

> original has 1-bit alpha channel, the output does not ...

Does that mean the output has no alpha, or 8-bit alpha, or what?

**Reply by Oblomov** [reply]
It has alpha, but it's  a single assigned color. I'm afraid I don't have the freedom to share the original image, and since the difficulty is in generating the exact image …

In the mean time, browsing the forum I have found [this thread](https://github.com/ImageMagick/ImageMagick/discussions/5475) from which I got the idea of using `-define png:color-type=6`, that gives an image with the same property as the original one, although ImageMagick  outpus a bKGD and cHRM chunk that were not in the original image.

**Comment by snibgo** [comment, marked as answer]
See https://imagemagick.org/script/defines.php for other defines that might help, such as "`png:exclude-chunk=value`".

**Comment by Oblomov** [comment]
Thank you! It would be nice if there was a way to tell IM directly “use the same as the original”, but with this I can at least achieve the same result manually.
